rr_supervisor/src/supervisor.cpp

```diff
@@ -1,16 +1,168 @@
+{
+    raceStarted=false;
+    twist_msg_count_ = 0;
+
+    null_vector_.x = 0.0; null_vector_.y = 0.0; null_vector_.z = 0.0;
+    twist_msg_.linear = null_vector_;
+    twist_msg_.angular = null_vector_;
+
+      // Setup publishers for twist multiplexer
+    twist_pub_ = nh_.advertise<geometry_msgs::Twist>(cmd, 1);
+    null_lock_ = nh_.advertise<std_msgs::Bool>(no_movement, 1);
+
+    // Setup subscribers to monitor battery and speed
+    cmd_sub_ = nh_.subscribe(vel_cmd, 1, &Supervisor::TrackSpeed, this);
+    battery_sub_ = nh_.subscribe(battery_state, 1, &Supervisor::MonitorBattery, this);
+
+    // Setup service servers
+    start_race_service_ = nh_.advertiseService(start_race, &Supervisor::StartRace, this);
+    count_lap_service_  = nh_.advertiseService(count_lap, &Supervisor::CountLap, this);
+
+}
+
+Supervisor::~Supervisor() 
+{
+    raceStarted=false;
+    twist_msg_count_ = 0;
+    null_vector_.x = 0.0; null_vector_.y = 0.0; null_vector_.z = 0.0;
+    twist_msg_.linear = null_vector_;
+    twist_msg_.angular = null_vector_;
+}
+
+// other methods
+void Supervisor::IdleRobot()
+{
+    ROS_INFO(¨Robot Idle!¨);
+    //Set null_lock as true. Will NOT allow path planner and joy messages through
+    bool_msg_.data=true;
+    null_lock_.publish(bool_msg_);
+}
+
+// Callback methods
+bool Supervisor::StartRace(std_srvs::Empty::Request &req, std_srvs::Empty::Response &res)
+{
+    raceStarted=true;
+    //Set null_lock as false. Will allow path planner and joy messages through
+    bool_msg_.data=false;
+    null_lock_.publish(bool_msg_);
+
+    begin_time =clock();
+    ROS_INFO(¨Race started!¨);
+    return true;
 }
 
-Supervisor::~Supervisor() {
+bool Supervisor::CountLap(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &res)
+{
+    if(race_type==¨drag¨)
+    {
+        lap_count_+=1;
+        res.success=true;
+        ROS_INFO(¨Drag race complete! shutting down¨)
+    }
+    else if(race_type_==¨circuit¨)
+    {
+        if (lap_count_ < 3)
+        {
+            lap_count_ += 1;
+            if (lap_count_ == 3)
+            {
+                res.success = true;
+                ROS_INFO("Circuit race complete! shutting down");
+            }
+            else
+            {
+                res.success = false;
+                ROS_INFO("Lap %d of %d complete!", lap_count_, 3);
+            }
+        }
+    }
+
+    //indicate successful run of the service 
+    if (res.success)
+    {
+        this->FinishRace();
+    }
+    return true;
+}
+void Supervisor::TrackSpeed(const geometry_msgs::TwistConstPtr& msg)
+{
+    twist_msg_count_+=1;
+    speed_sum_+=(pow(msg->linear.x, 2) + pow(msg->linear.y, 2))
+}
+
+void Supervisor::MonitorBattery(const std_msgs::Int8::ConstPtr& msg)
+{
+  if (msg->data <= 10)
+  {
+    ROS_INFO("Battery is less than 10 percent, ending race...");
+    this->FinishRace();
+  }
 }
+
+void Supervisor::FinishRace()
+{
+  // Calculate average speed and elapsed time since start of race
```

Comment:
Honestly I think all these race metrics logging isn't really worth it. I kinda added it last year as a 'nice to have' feature but it was ultimately not used at all and I find it hard to imagine this information will be of value to someone in the future. Any thoughts? @JXproject @DongJunJin